fathom/src/symbol.rs

```diff
@@ -0,0 +1,174 @@
+    /// // ...
+    /// assert_eq!(Symbol::get_alphabetic_name(51), Symbol::intern("z1"));
+    /// assert_eq!(Symbol::get_alphabetic_name(52), Symbol::intern("a2"));
+    /// // ...
+    /// ```
+    pub fn get_alphabetic_name(&mut self, index: usize) -> Symbol {
+        self.reserve_alphabetic_names(index);
+        self.alphabetic_names[index]
+    }
+
+    /// Allocate and intern all tuple labels up-to and including `max_index`
+    /// if they are not already present.
+    pub fn reserve_tuple_labels(&mut self, max_index: usize) {
+        fill_vec(&mut self.tuple_labels, max_index, |index| {
+            Symbol(self.strings.get_or_intern(format!("_{index}")))
+        })
+    }
+
+    /// Get or intern a string in the form `_{index}`.
+    ///
+    /// ## Example
+    ///
+    /// ```rust
+    /// use fathom::symbol::Symbol;
+    ///
+    /// assert_eq!(Symbol::get_tuple_label(0), Symbol::intern("_0"));
+    /// assert_eq!(Symbol::get_tuple_label(1), Symbol::intern("_1"));
+    /// ```
+    pub fn get_tuple_label(&mut self, index: usize) -> Symbol {
+        self.reserve_tuple_labels(index);
+        self.tuple_labels[index]
+    }
+
+    /// Get or intern a slice of strings in the form `_{index}` for each index
+    /// in `range`.
+    pub fn get_tuple_labels(&mut self, range: Range<usize>) -> &[Symbol] {
+        self.reserve_tuple_labels(range.end.saturating_sub(1));
+        &self.tuple_labels[range]
+    }
+
+    /// Returns true if `label` refers to a string in the form `_{index}`.
+    pub fn is_tuple_label(&mut self, index: usize, label: Symbol) -> bool {
+        label == self.get_tuple_label(index)
+    }
+
+    /// Returns true if `labels` is a sequence of tuple labels: `_0`, `_1`, ...
+    pub fn is_tuple_labels(&mut self, labels: &[Symbol]) -> bool {
+        labels == self.get_tuple_labels(0..labels.len())
+    }
+}
+
+fn alphabetic_name(index: usize) -> String {
+    let base = index / 26;
+    let letter = index % 26;
+    let letter = (letter as u8 + b'a') as char;
+    if base == 0 {
+        format!("{letter}")
+    } else {
+        format!("{letter}{base}")
+    }
+}
+
+fn fill_vec<T>(vec: &mut Vec<T>, max_index: usize, f: impl FnMut(usize) -> T) {
+    vec.extend((vec.len()..=max_index).map(f))
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Symbol(lasso::Spur);
+
+impl Symbol {
+    pub fn intern(sym: impl AsRef<str>) -> Self {
+        Self(INTERNER.write().unwrap().strings.get_or_intern(sym))
+    }
+
+    pub fn intern_static(sym: &'static str) -> Self {
+        Self(INTERNER.write().unwrap().strings.get_or_intern_static(sym))
+    }
+
+    pub fn resolve(self) -> &'static str {
+        let interner = INTERNER.write().unwrap();
+        let symbol = interner.strings.resolve(&self.0);
+
+        // SAFETY: `INTERNER` is static, so references into it are `'static`
+        unsafe { std::mem::transmute::<&str, &str>(symbol) }
+    }
```

Comment:
Hmm, how sure are we that this is safe and not UB? Can we run it by MIRI? I would at least prefer we used `std::mem::transmute::<&str, &'static str>(symbol)` to be explicit.

Reply:
I guess I’ve always associated `'static` with stuff that is in the static code of the program, where as the strings in `ThreadedRodeo` are being allocated at runtime. I could be off on this understanding, however. It does give me pause though.

Reply:
My understanding is `'static` is simply the lifetime that outlives all others. Usually comes from string literals, but can also come from eg leaking an allocation.

Reply:
If you are uncomfortable with the `'static` lifetime, we could change the signature to 
`fn resolve<'a>(&'a self) -> &'a str` (with explicit lifetimes added for clarity).

[Taken from rustc](https://github.com/rust-lang/rust/blob/e4dd9edb76a34ecbca539967f9662b8c0cc9c7fb/compiler/rustc_span/src/symbol.rs#L1845):

```rs
impl Symbol {
    /// Access the underlying string. This is a slowish operation because it
    /// requires locking the symbol interner.
    ///
    /// Note that the lifetime of the return value is a lie. It's not the same
    /// as `&self`, but actually tied to the lifetime of the underlying
    /// interner. Interners are long-lived, and there are very few of them, and
    /// this function is typically used for short-lived things, so in practice
    /// it works out ok.
    pub fn as_str(&self) -> &str {
        with_session_globals(|session_globals| unsafe {
            std::mem::transmute::<&str, &str>(session_globals.symbol_interner.get(*self))
        })
    }
}
```

Reply:
Ahh, fair. If it’s ok for rustc then it’s ok for me! Let’s change the type signature.